src/components/SelectionController.tsx

```diff
@@ -0,0 +1,292 @@
+import { useEffect, useReducer, useCallback } from 'react';
+import {
+  Selection,
+  FileData,
+  InputEvent,
+  SelectionType,
+  FileArray,
+} from '../typedef';
+import { Nullable } from 'tsdef';
+import { isArray, isNumber } from 'util';
+import {
+  clampIndex,
+  isObject,
+  isNil,
+  isMobileDevice,
+  isFunction,
+} from '../util/Util';
+import ConsoleUtil from '../util/ConsoleUtil';
+import { KbKey, Options, SortOrder } from '..';
+import { usePrevious } from '../hooks/usePrevious';
+import { shallowEqualArrays } from 'shallow-equal';
+
+interface RenderProps {
+  previousSelectionIndex?: number;
+  selection: Selection;
+  handleFileSingleClick: (
+    file: FileData,
+    event: InputEvent,
+    displayIndex: number
+  ) => void;
+  setSelection: (
+    selection: Selection,
+    selectionIndex: number | undefined
+  ) => void;
+}
+
+interface Props {
+  children: (props: RenderProps) => any;
+  files?: Nullable<FileData>[];
+  disableSelection: boolean;
+  folderChain: FileArray;
+  onSelectionChange?: (selection: Selection) => void;
+  options?: Options;
+  sortProperty?: string | ((file: FileData) => any);
+  sortOrder?: SortOrder;
+}
+
+type State = Pick<RenderProps, 'selection' | 'previousSelectionIndex'>;
+
+interface FileSingleClickAction {
+  type: 'file_single_click';
+  files?: Nullable<FileData>[];
+  file: FileData;
+  displayIndex: number;
+  event: InputEvent;
+}
+
+type Action =
+  | {
+      type: 'new_files';
+      files?: Nullable<FileData>[];
+    }
+  | { type: 'new_folder_chain' }
+  | { type: 'disable_selection' }
+  | {
+      type: 'set_selection';
+      selection: Selection | undefined;
+      selectionIndex: number | undefined;
+    }
+  | {
+      type: 'new_sort';
+      files?: Nullable<FileData>[];
+    }
+  | FileSingleClickAction
+  | { type: 'key_press' };
```

Comment:
Can you please put the action types here and below into a separate enum?

Reply:
Yes, how would you like them cased?  `ActionTypes.FileSingleClick`?

Reply:
Yeah, that works.